### PR TITLE
bug/1749_list_of_types_in_query

### PR DESCRIPTION
**Pull request by kzangeli**
Implemented lists of types in URI param **type**.
Fixes issue #1749

### CODE REVIEW

File: src/lib/serviceRoutinesV2/getEntities.cpp [hunk 1]
```diff
@@ -229,6 +234,35 @@ std::string getEntities
...
+    parseDataP->qcr.res.fill(pattern, "", "true", EntityTypeEmptyOrNotEmpty, "");
+  }
+  else if (ciP->uriParamTypes.size() == 1)
+  {
+    parseDataP->qcr.res.fill(pattern, ciP->uriParam["type"], "true", EntityTypeNotEmpty, "");
+  }
+  else
+  {
+    //
+    // More than one type listed in 'q':
```
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Type is not part of q=... I mean q= is one URL query string parameter and type= is another. Thus, the comment could be missleanding.

File: src/lib/serviceRoutinesV2/getEntities.cpp [hunk 2]
```diff
@@ -81,6 +83,7 @@ std::string getEntities
   std::string  geometry   = ciP->uriParam["geometry"];
   std::string  coords     = ciP->uriParam["coords"];
   std::string  out;
+  bool         isPattern  = true;
```
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Not sure why this is needed. Note that, no matter if id= or idPattern= is used, mongoBackend needs always patterns (note how the entity id is built, even in the case of a single id= element).

File: src/lib/serviceRoutinesV2/getEntities.cpp [hunk 3]
```diff
@@ -229,6 +234,35 @@ std::string getEntities
...
+  }
+  else
+  {
+    //
+    // More than one type listed in 'q':
+    // Add an entity per type to QueryContextRequest::entityIdVector
+    //
+    for (unsigned int ix = 0; ix < ciP->uriParamTypes.size(); ++ix)
+    {
+      EntityId* entityId = new EntityId(pattern, ciP->uriParamTypes[ix], isPattern? "true" : "false");
```
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Related with my comment above, I think this should use "true" isPattern.

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+--NAME--
+V2 query with array in URI param 'type'
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
```
Comment (fgalan) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: This .test used type= "standalone".

To get a better case coverage, I'd suggest to add additional test using at the same time id= and type= and idPattern= and type=.

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Of course, not 'q' but 'type'.
cut 'n paste ...

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Yes, I understood. Always "true"

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: This PR is not really about id and idPattern, however, as tests seem to be missingm, let's add some ...

Comment (kzangeli) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: Yes, agreed. More tests needed.

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Fixed in 6e9b5bf

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Fixed in 6e9b5bf

File: src/lib/serviceRoutinesV2/getEntities.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntities.cpp]: Fixed in 6e9b5bf

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: Fixed in 631120f

File: CHANGES_NEXT_RELEASE [hunk 5]
```diff
@@ -1,3 +1,4 @@
-- Fix: libmicrohttpd 0.9.48 included in contextBroker as static lib (previous Orion versions used 0.99.22 as dynamic library) (Issue #1675)
-- Fix: error level traces ignoring -logLevel NONE
-- Fix: wrong over-logging at error level updating attributes having metadata without type 
+-Fix: error level traces ignoring -logLevel NONE
+-Fix: wrong over-logging at error level updating attributes having metadata without type 
+-Fix: libmicrohttpd 0.9.48 included in contextBroker as static lib (previous Orion versions used 0.99.22 as dynamic library) (Issue #1675)
+-Fix: list of attribute names in URI param 'type' (Issue #1749)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Looking to the last CNR version before the last cleanup (https://github.com/telefonicaid/fiware-orion/blob/a128b4c715c7e7d1881cc5614b13390f2dd6871e/CHANGES_NEXT_RELEASE) the "style" is to put one whitespace between "-" and "Fix".

(I have the feeling that each time CNR is flused each time we open release, we "flip" to another style variant :))

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+#
+# 01. Create an entity E1 of type T1
+# 02. Create an entity E1 of type T2
+# 03. Create an entity E1 of type T3
+# 04. GET all entities (3 entities)
+# 05. GET /v2/entities?type=T1, see E1/T1
+# 06. GET /v2/entities?type=T1,T2 see E1/T1 and E1/T2
+# 07. GET /v2/entities?type=T1,T3 see E1/T1 and E1/T3
+# 08. GET /v2/entities?type=T1,T2,T3 see E1/T1, E1/T2 and E1/T3
+# 09. GET /v2/entities?type=T1,T2,T3&id=E1, see E1/T1, E1/T2 and E1/T3
```
Comment (fgalan) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: Maybe we should add a case with id= with enumeration, i.e. id=E1,E2?

Comment (kzangeli) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: I dind't even know that existed ... :-)

File: CHANGES_NEXT_RELEASE (hunk 5 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: I had a space between '-' and 'Fix', but merging with develop left it that way ...
Changing it back

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: Fixed in ce91f42

File: CHANGES_NEXT_RELEASE (hunk 5 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in ce91f42

File: test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1749_list_of_types_in_query/list_of_types_in_query.test]: Don't underestimate the power of the pull request as a learning tool :D

**Comment by fgalan** [comment]
LGTM (passing the ball to @crbrox for final LGMT + merge + delete branch)

**Comment by crbrox** [comment]
LGTM

**Pull request merged**